src/dbnode/encoding/arrow/tile/counter_resets_downsampler.go

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package tile
+
+import (
+	"math"
+)
+
+// DownsampledValue represents a datapoint value after downsampling.
+type DownsampledValue struct {
+	// FrameIndex is the index to the original datapoint within source frame (before downsampling).
+	FrameIndex int
+	// Value is a downsampled datapoint value.
+	Value float64
+}
+
+// DownsampleCounterResets downsamples datapoints in a way that preserves counter invariants:
+// The the last (and the first, if necessary) values stay the same as original (to transfer counter state from tile to tile).
+// Also, after applying counter reset adjustment logics, all the values would be the same as
+// after applying this logics to the original values.
+// As an optimization (to reduce the amount of datapoints), prevFrameLastValue can be passed in (if available),
+// and then in some cases the first value of this tile may be omitted.
```

Comment:
Any arguments for always having this additional first value per tile specifically on counters (as opposed to gauges)?
I would prefer to avoid this, as this would double the number of datapoints after downsampling for a counter that has no resets.

Reply:
One more thing to take into consideration here - always keeping both the first and the last datapoint of each frame would not only double the amount of datapoints after downsampling, but also would hurt the delta-of-delta timestamp compression rate, as the datapoints would be unequally spaced.